### PR TITLE
Mvc inversify repo branch

### PR DESCRIPTION
**Pull request by YSherbini**

### CODE REVIEW

File: src/repositories/CategoryRepository.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { injectable } from 'inversify';
+import Category from '../models/category.js';
+
+@injectable()
+export class CategoryRepository {
+    async getAll() {
+        try {
```
Comment (mismailk) [inline src/repositories/CategoryRepository.ts]: No need for try and catch

File: src/controllers/CategoryController.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import auth from '../middleware/auth.js';
+import { isValidObjectId } from '../middleware/validate.js';
+
+@controller('/categories')
+export class CategoryController {
+    constructor(@inject(CategoryRepository) private readonly categoryRepository: CategoryRepository) {}
+
+    @httpGet('/', auth)
+    async allCategories(req: IRequest, res: express.Response) {
+        try {
```
Comment (mismailk) [inline src/controllers/CategoryController.ts]: No need for try catch

File: src/controllers/CategoryController.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import express from 'express';
+import { controller, httpGet } from 'inversify-express-utils';
+import { CategoryRepository } from '../repositories/CategoryRepository.js';
+import { inject } from 'inversify';
+import { IRequest } from '../interfaces/IRequest.js';
+import auth from '../middleware/auth.js';
+import { isValidObjectId } from '../middleware/validate.js';
+
+@controller('/categories')
```
Comment (mismailk) [inline src/controllers/CategoryController.ts]: You can add auth middleware within @controller to be applied for all routes in the class

File: src/interfaces/IQuery.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import { ICategory } from "./ICategory.js";
+import { IEvent } from "./IEvent.js";
+
+export interface IQuery {
```
Comment (mismailk) [inline src/interfaces/IQuery.ts]: Choose a meaningfull name for the interface as EventParams or EventQueryParams.

File: src/repositories/EventRepository.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+            return await Event.findById({ _id }).populate(populate, '-__v');
+        } catch (err) {
+            throw new Error('Coudnt get event');
+        }
+    }
+
+    applyQuery(eventQuery: IQuery) {
+        const match: any = {}
+        const { name, categoryId } = eventQuery
+        if (typeof name === 'string') {
```
Comment (mismailk) [inline src/repositories/EventRepository.ts]: What is the need for type checking you already define the type on IQuery interface

File: src/repositories/RegisterRepository.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                },
+                {
+                    $lookup: {
+                        from: 'events',
+                        localField: 'eventId',
+                        foreignField: '_id',
+                        as: 'eventDetails',
+                    },
+                },
+                {
```
Comment (mismailk) [inline src/repositories/RegisterRepository.ts:42]: use $unwind to convert an array with one element to an object

File: src/controllers/TeamController.ts [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        const gymOwnerId = req.gymOwner._id;
+        try {
+            const register = await this.registerRepository.getOne({ _id: registerId, gymOwnerId }, 'event teams');
+            if (!register) {
+                return res.status(404).send('Register not found!');
+            }
+            if (register.event && !register.event.categoriesIds.includes(categoryId)) {
+                return res.status(400).send('Categories doesnt match!');
+            }
+            let team = await this.teamRepository.create({ name, categoryId, registerId, gymOwnerId, eventId: register.eventId });
```
Comment (mismailk) [inline src/controllers/TeamController.ts]: You may receive eventId as a parameter in the request body

File: src/repositories/TeamRepository.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+                    return dupTeams;
+                }
+                if (register.event && !register.event.categoriesIds.includes(team.categoryId)) {
+                    dupTeams.error = { status: 422, msg: "Categories does't match" };
+                    return dupTeams;
+                }
+                const { name, image, categoryId, gymOwnerId, registerId, eventId } = team;
+                const teamParams: ITeamParams = { name, image, categoryId, gymOwnerId, registerId, eventId };
+                const dupTeam = await this.create(teamParams);
+                dupTeams.teams.push(dupTeam);
```
Comment (mismailk) [inline src/repositories/TeamRepository.ts]: You can simplify this more using object destructuring.

```suggestion
                const dupTeam = await this.create({...team.toObject(), _id: undefined});
                dupTeams.teams.push(dupTeam);
```

File: src/repositories/TeamRepository.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
+import { injectable } from 'inversify';
+import Team from '../models/team.js';
+import { IDuplicateRes, ITeam, ITeamParams } from '../interfaces/ITeam.js';
+import { IQuery } from '../interfaces/IQuery.js';
+import { IRegister } from 'interfaces/IRegister.js';
+
+@injectable()
+export class TeamRepository {
+    async create(teamParams: ITeamParams) {
+        return new Team(teamParams);
```
Comment (mismailk) [inline src/repositories/TeamRepository.ts:16]: This isn't an async operation

File: src/repositories/TeamRepository.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+            return await team.save();
+        } catch (err) {
+            throw new Error('Coudnt save!');
+        }
+    }
+
+    async saveAll(teams: ITeam[]) {
+        for (const team of teams) {
+            await this.save(team);
+        }
```
Comment (mismailk) [inline src/repositories/TeamRepository.ts]: You can use Promise.all to run an asynchronous in parallel.

File: src/repositories/TeamRepository.ts [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        const sort: Record<string, number> = {};
+        if (categoryId) {
+            match.$and.push({ categoryId });
+        }
+        if (categoriesIds.length > 0) {
+            match.$and.push({ categoryId: { $in: categoriesIds } });
+        }
+        if (sortBy) {
+            const [sortee, order] = sortBy.split(':');
+            sort[sortee] = order === 'desc' ? -1 : 1;
```
Comment (mismailk) [inline src/repositories/TeamRepository.ts]: Don't use magic numbers or strings define an enum to inhance redability.

**Comment by mismailk** [comment]
1. Add more spaces between lines to make it more readable.
2. Remove unnecessary try catches blocks.

**Review by mismailk — CHANGES_REQUESTED** [review]

**Comment by YSherbini** [comment]
I've implemented the requested changes and changed code to commonjs.

**Pull request closed without merging**